### DISCUSSION TITLE
Smart Meter Interface/Scripting/UFS/MQTT: TLS-MQTT connection fails, error -1000 (Out of Memory Error)

### DISCUSSION THREAD

**Discussion by GolterU** [arendst/Tasmota#19363] [Q&A]
I am using an [ESP-01S](https://templates.blakadder.com/ESP-01S.html) with Tasmota-DE to read the data from two smart meters successfully. The connection to the [Cloud HiveMQ MQTT broker](https://console.hivemq.cloud/?utm_source=HiveMQ+Cloud+Broker+Page&utm_medium=signup+free+topleft+CTA+Button&utm_campaign=HiveMQ+Cloud+PaaS&utm_content=serverless) with TLS also works.
However, the same configuration with an [ESP-12F](https://templates.blakadder.com/ESP-12.html) will not work. In addition to the other hardware, the difference is the use of UFS (`#define USE_UFILESYS` and `board = esp8266_4M2M` in platformio_override.ini).
As soon as MQTT is active on the ESP-12F (setOption3 true), the ESP-12F can only be reached sporadically via the WLAN. Luckily, the console is the most visible (main menu, script console, etc. mostly not). The console reports:
```
00:00:00.001 HDW: ESP8266EX
00:00:00.025 UFS: FlashFS mounted with 1952 kB free
00:00:00.092 CFG: Loaded from File, Anzahl 131
00:00:00.139 QPC: Count 1
00:00:00.160 Projekt tasmota - Tasmota Version 13.0.0(tasmota)-2_7_4_9(2023.08.06 15:28:13)
00:00:00.161 UFILESYSTEM OK!
00:00:00.769 WIF: verbinden mit AP1 HUETTENFUNK1 Channel 1 BSSId E8:DF:70:74:xx:xx in Modus 11n als ESP-12F-1...
00:00:01.757 WIF: verbunden
00:00:02.009 HTP: Web-Server aktiv bei ESP-12F-1 mit IP-Adresse 192.168.0.121
21:54:32.024 MQT: Verbindungsversuch...
21:54:32.243 MQT: TLS connection error: -1000
21:54:32.245 MQT: Verbindung fehlgeschlagen aufgrund von MyBroker.s2.eu.hivemq.cloud:8883, rc -2. Wiederversuch in 10 s
21:54:35.255 QPC: Reset
21:54:43.039 MQT: Verbindungsversuch...
21:54:43.081 MQT: TLS connection error: -1000
21:54:43.083 MQT: Verbindung fehlgeschlagen aufgrund von MyBroker.s2.eu.hivemq.cloud:8883, rc -2. Wiederversuch  20 s
...
```
The memory configuration looks like this:
**ESP-01S**
- Flash chip size 1024 KB
- Flash usable 1024 KB
- Size program 742 KB
- Flash free 260 KB
- Free memory 28.9 KB

**ESP-12F** (with UFS)
- Flash chip size 4096 KB
- Flash usable 4096 KB
- Size program 814 KB
- Flash free 1232 KB
- Free memory 11.8 KB
- File system size 2024 KB

After attempting to reduce the firmware with the following #undef:
USE_EMULATION_HUE
USE_EMULATION_WEMO
USE_TELEGRAM
USE_IR_REMOTE
USE_IR_RECEIVE
USE_DOMOTICZ
USE_I2C

This increases the available RAM to 14.4 kByte, but does not change the -1000 error.

What else can I do?

Regards
Golter

**Comment by sfromis** [comment]
You could run the tool decode-config to get a listing of features enabled, to see what more you can exclude. However, *most* features do not use much RAM until active. Scripting also has a number of settings `#define` values you could consider trimming.

More generally, with ESP8266 it is not really workable to just add features on top of each other. You can get a lot more headroom using a  device based on ESP32.

**Comment by GolterU** [comment]
Hi @sfromis,
Thanks for your hints.
I tested the decode-config tool. Unfortunately, I don't understand how the possible or unnecessary compiler directives can be recognized from the results.
For example, despite `#undef USE_I2C`, decode-config shows tons of related entries. On the other hand, there is no relevant entry in output.json for `#define USE_SCRIPT_WEB_DISPLAY`.

Maybe an ESP32 really is the best solution.

Kind regards
Golter

**Reply by sfromis** [reply]
Hmmm. What I really meant was not decode-config but instead decode-status, where there is a clear list of included features.

The tool decode-config shows all available details, including for drivers not included, when they still have reserved flags for configuration.

**Comment by GolterU** [comment]
Hi @sfromis,
Thank you for your further information.
The decoded information gives you a little idea of what a huge project Tasmota is dealing with. All respect.

Kind regards
Golter
